types/validator_set.go

```diff
@@ -569,22 +557,44 @@ func (vals *ValidatorSet) updateWithChangeSet(changes []*Validator, allowDeletes
 
 	// Compute the priorities for updates
 	numNewValidators := computeNewPriorities(updates, vals, updatedTotalVotingPower)
-	if len(vals.Validators)+numNewValidators <= len(deletes) {
-		err = fmt.Errorf("applying the validator changes would result in empty set")
-		return err
+	if len(vals.Validators)+numNewValidators == len(deletes) {
+		return errors.New("applying the validator changes would result in empty set")
+	}
+
+	if len(vals.Validators)+numNewValidators < len(deletes) {
```

Comment:
Does it make sense to include the new validators here, too? E.g., does it make sense to add new bunch of new validators and remove all those again in the same operation? Are deletes not only for the existing vals?

Reply:
This basically checks for negative length after applying everything but I agree that it does look weird so I removed it. We do verify earlier that deletes are only for existing validators.
There was a panic (for len(deletes)>len(vals.Validators)) in applyRemovals that I just moved it in verifyRemovals.